<a href="https://colab.research.google.com/github/htapiagroup/sistemas-distribuidos-evaluacion-intersemestral-cflores23/blob/master/Sist_Dist2_Examen1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Primer examen Parcial

Alumno: Camilo Flores Rodríguez

Trabajé solo

#Importar datos
      1. Adquirir datos
      Para adquirir datos descargo con la instrucción del S.O. !wget
      
      He utilizado mi Google drive por lo que debo motar la unidad

In [0]:
!wget https://archive.org/download/nycTaxiTripData2013/trip_data.7z

In [28]:
from google.colab import drive
drive.mount('/content/gdrive')


Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


Observo donde me encuentro ubicado:

In [0]:
ls

gdrive/  sample_data/


In [0]:
ls


gdrive/  sample_data/


Puedo observar que sí se descargó en mi unidad drive los archivos que necesito para el ejercicio.

Como ya he montado mi unidad de Google drive me cambio a la carpeta Colab Notebooks

In [0]:
cd gdrive/My\ Drive/'Colab Notebooks'

/content/gdrive/My Drive/Colab Notebooks


In [0]:
Para poder responder la pregunta 3 necesito utilizar la liberría Pandas por lo que la imoporto de una vez

In [0]:
import pandas as pd, time

3.  Una vez descargado el archivo pueden descomprimirlo usando la siguiente instruccion

Continúo con descomprimir un archivo

In [0]:
!7z e trip_data.7z -o/content/gdrive/My\ Drive/Colab\ Notebooks/ExamData trip_data_1.csv


7-Zip [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.UTF-8,Utf16=on,HugeFiles=on,64 bits,2 CPUs Intel(R) Xeon(R) CPU @ 2.30GHz (306F0),ASM,AES-NI)

Scanning the drive for archives:
  0M Scan         1 file, 4102781969 bytes (3913 MiB)

Extracting archive: trip_data.7z
--
Path = trip_data.7z
Type = 7z
Physical Size = 4102781969
Headers Size = 433
Method = LZMA:24
Solid = -
Blocks = 12

  0%    
Would you like to replace the existing file:
  Path:     /content/gdrive/My Drive/Colab Notebooks/ExamData/trip_data_1.csv
  Size:     2459600863 bytes (2346 MiB)
  Modified: 2014-05-12 22:02:53
with the file from archive:
  Path:     trip_data_1.csv
  Size:     2459600863 bytes (2346 MiB)
  Modified: 2014-05-12 22:02:53
? (Y)es / (N)o / (A)lways / (S)kip all / A(u)to rename all / (Q)uit? Y

  0% - trip_data_1.csv                        1% - trip_data_1.csv

Busco donde se descomprimió el archivo

In [0]:
ls

'Copia de Copia de SisDist_S1_L1.ipynb'                Examen_Final.ipynb
'Copia de Copy of MCA_SistDist_Ex1'                    Sesion2_Sist_dist2.ipynb
'Copia de Hola, Colaboratory'                          SisDist2_Ex1.ipynb
'Copia de Hola, Colaboratory (1)'                     'SisDist2_S2 (1).ipynb'
'Copia de Hola, Colaboratory (2)'                      SisDist2_S2.ipynb
'Copia de Sesion2_Sist_dist2.ipynb'                    SisDist2_sesion2.ipynb
'Copia de SisDist_S1_L1.ipynb'                         SisDist_S1_L1.ipynb
'Copia de SisDist_S5.ipynb'                            SisDist_S5.ipynb
'Copia de SistDist_S4 (1).ipynb'                       SistDist2/
'Copia de SistDist_S4.ipynb'                           SistDist2_Exam1.ipynb
'Copia de SistDist_S6'                                 Sist_Dist2_Examen1.ipynb
'Copia de Te damos la bienvenida a Colaboratory'       SistDist_S4.ipynb
'Copia de Te damos la bienvenida a Colaboratory (1)'   SistDist_S6.ipynb
'Copia de Te damos la bienv

El archivo se encuentra en la carpeta que cree ExamData, por lo tanto me cambio de directorio

In [0]:
cd ExamData

/content/gdrive/My Drive/Colab Notebooks/ExamData


In [0]:
ls

trip_data_1.csv  Untitled0.ipynb


Puedo observar de la celda anterior que el archivo trip_data_1.csv si se encuentra en la carpeta

Leo el archivo con pandas

In [0]:
df = pd.read_csv('trip_data_1.csv')

Verifico si lo leyó bien, para ello visualizo los primeros datos con los nombres de las columnas

In [0]:
df.head()

medallion  ... dropoff_latitude
0  89D227B655E5C82AECF13C3F540D4CF4  ...        40.751171
1  0BD7C8F5BA12B88E0B67BED28BEA73D8  ...        40.750660
2  0BD7C8F5BA12B88E0B67BED28BEA73D8  ...        40.726002
3  DFD2202EE08F7A8DC9A57B02ACB81FE2  ...        40.759388
4  DFD2202EE08F7A8DC9A57B02ACB81FE2  ...        40.747868

[5 rows x 14 columns]

Respondo a las preguntas 

3. Responder las siguientes preguntas:

    1.  Cuantas columnas contiene cada archivo de datos descomprimido
    2.  Cuantos renglones tiene cada archivo
    
    Para ello utilizo shape de pandas y me da la respuesta que busco. Procedo a leer cada archivo para responder las preguntas con todos los archivos, debido a que hice el primer ejercicio con un solo archivo.

In [0]:
df.shape

(14776615, 14)

In [0]:
boleana = pd.isnull(df)
print (boleana)

De acuerdo a la consulta anterior tenemos 14 columnas y 14776615 líneas o registros

4. Identificar si existen renglones con errores en los datos, por ejemplo si hay columnas de mas (o de menos), si hay campos vacios, etc.

    1. indicar claramente cuantos son los renglones con errores y a que archivos corresponden

In [0]:
plantilla = "trip_data_{}.csv"
for x in range(1, 4):
  df_temp = da.read_csv(plantilla.format(x), dtype={' store_and_fwd_flag': 'object'})
  print(df_temp.compute().shape)

(14776615, 14)


/usr/local/lib/python3.6/dist-packages/dask/core.py:118: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  args2 = [_execute_task(a, cache) for a in args]


(13990176, 14)
(15749228, 14)


In [0]:
nulos=len(df[pd.isnull(df.medallion)])
print(nulos)

0


Para responder si hay o no valores nulos esta última instrucción nos sirve, solo que debemos iterarla para que lea todas las columnas.

Voy a crear una lista con los nombres de las columnas del archivo y despues recorreré la lista con un for, para determinar si tienen o no valores nulos

#Columnas con valores nulos:

In [0]:
nombres_cols = df.columns.tolist()

for column in nombres_cols:
  print("Valores nulos en la columna <{0}>: {1}".format(column, df[column].isnull().sum()))


Valores nulos en la columna <medallion>: 0
Valores nulos en la columna <hack_license>: 0
Valores nulos en la columna <vendor_id>: 0
Valores nulos en la columna <rate_code>: 0
Valores nulos en la columna <store_and_fwd_flag>: 7326207
Valores nulos en la columna <pickup_datetime>: 0
Valores nulos en la columna <dropoff_datetime>: 0
Valores nulos en la columna <passenger_count>: 0
Valores nulos en la columna <trip_time_in_secs>: 0
Valores nulos en la columna <trip_distance>: 0
Valores nulos en la columna <pickup_longitude>: 0
Valores nulos en la columna <pickup_latitude>: 0
Valores nulos en la columna <dropoff_longitude>: 86
Valores nulos en la columna <dropoff_latitude>: 86


Podemos observar que nuestro archivo si tiene valores nulos. 

Ahora haremos lo mismo pero con los otros archivos

In [0]:
nombres_cols = df_temp .columns.tolist()
for x in range(1, 4):
  archivo = pd.read_csv(plantilla.format(x))
  for column in nombres_cols:
    print("Valores nulos en la columna <{0}>: {1}".format(column, df_temp[column].isnull().sum()))
  print("**************************************************************")

Valores nulos en la columna <medallion>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < hack_license>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < vendor_id>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < rate_code>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < store_and_fwd_flag>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < pickup_datetime>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < dropoff_datetime>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < passenger_count>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < trip_time_in_secs>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < trip_distance>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < pickup_longitude>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < pickup_latitude>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la c

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Valores nulos en la columna <medallion>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < hack_license>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < vendor_id>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < rate_code>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < store_and_fwd_flag>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < pickup_datetime>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < dropoff_datetime>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < passenger_count>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < trip_time_in_secs>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < trip_distance>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < pickup_longitude>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la columna < pickup_latitude>: dd.Scalar<series-..., dtype=int64>
Valores nulos en la c

     2. crear un nuevo conjunto de datos con los errores eliminados

In [0]:
df_sinNulos = df.dropna(how='any')

Ahora volvamos a correr la instrucción de la celda 28 para verificar si se eliminaron los nulos

In [0]:
nombres_cols = df_sinNulos.columns.tolist()

for column in nombres_cols:
  print("Valores nulos en la columna <{0}>: {1}".format(column, df_sinNulos[column].isnull().sum()))

Valores nulos en la columna <medallion>: 0
Valores nulos en la columna <hack_license>: 0
Valores nulos en la columna <vendor_id>: 0
Valores nulos en la columna <rate_code>: 0
Valores nulos en la columna <store_and_fwd_flag>: 0
Valores nulos en la columna <pickup_datetime>: 0
Valores nulos en la columna <dropoff_datetime>: 0
Valores nulos en la columna <passenger_count>: 0
Valores nulos en la columna <trip_time_in_secs>: 0
Valores nulos en la columna <trip_distance>: 0
Valores nulos en la columna <pickup_longitude>: 0
Valores nulos en la columna <pickup_latitude>: 0
Valores nulos en la columna <dropoff_longitude>: 0
Valores nulos en la columna <dropoff_latitude>: 0


Como podemos observar en el resultado anterior, hemos creado un nuevo conjunto llamado df_sinNulos sin los errores 

Para hacer el cómputo más rápido y eficiente dividiré el archivo en trozos.

Debido a que el archvo contiene 14776615 lineas lo dividiré primero en tamaños de 500,000 y probaremos si este tamaño es eficiente.

In [0]:
archivo = 'trip_data_1.csv'

#5. Crear un DataFrame usando la libreria pandas y responder lo siguiente:

    1. Indicar el tiempo en segundos, que tarda la libreria pandas en leer un archivo a un DataFrame
    2. Cual es la huella de memoria del proceso usado para generar el objeto DataFrame
    3. Cuanta memoria del sistema se usa para crear el objeto anterior
    4. Indicar el tiempo que tarda pandas en obtener el promedio de la distancia de viaje (trip_distance)

El dataframe ya se ha creado arriba para resolver las preguntas anteriores. Sin embargo, para fines de responder las del punto 5 lo volveremos a crear utilizando los mismos datos

In [0]:
t_start = time.time()
df = pd.read_csv('trip_data_1.csv')
t_end = time.time()
print('Tiempo de lectura: {} s'.format(t_end-t_start)) # time [s]

Definimos una función que nos permita medir la huella de memoria

In [0]:
import psutil, os

def huella_memoria():
  '''Regresa la huella de memoria usada por un proceso de Python'''
  mem = psutil.Process(os.getpid()).memory_info().rss
  return (mem / 1024**2)

In [0]:
antes = huella_memoria()

In [0]:
n = (1024**2)//8

In [0]:
df = pd.read_csv('trip_data_1.csv')

In [0]:
despues = huella_memoria()

Respondemos pregunta 2

In [0]:
print('Memoria antes: {} MB'.format(antes))

Memoria antes: 143.6328125 MB


In [0]:
print('Memoria despues: {} MB'.format(despues))

Memoria despues: 5107.66796875 MB


Para responder la pregunta 3 hacemos una resta para saber la diferencia entre la memoria anterior y la actual, así sabremos cuanta memoria ocupa el objeto.

In [0]:
mem_objeto = despues-antes

print('La cantidad de memoria que ocupa el dataframe es de: {} MB'.format(mem_objeto))

La cantidad de memoria que ocupa el dataframe es de: 4964.03515625 MB


In [0]:
promedio=df['trip_distance'].mean()
print(promedio)

2.770975670679966


In [0]:
for trozo in pd.read_csv(archivo, chunksize=500000):
  print('type: %s shape %s' % (type(trozo), trozo.shape))

type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (500000, 14)
type: <class 'pandas.core.frame.DataFrame'> shape (5000

6. Todos los pasos siguientes deben realizarse usando la libreria dask
    1. Indicar el tiempo que tarda en obtener el promedio de la distancia de viaje (trip_distance)
    2. Indicar el tiempo que tarda en obtener el promedio de la duracion de viaje (trip_time_in_secs)
    3. Con las columnas pickup_datetime y dropoff_datetime como objetos temporales, crear una nueva columna en los dataframes que se llame duracion y compararla renglon a renglon con la columna trip_time_in_secs
    4. Definir un viaje_largo como aquellos en los que la duracion del viaje (trip_time_in_secs) sea mayor a 20 minutos.

Para responder la pregunta imoprto las librerías dask y time

In [0]:
import dask.dataframe as da, time

Me regreso un nivel en la carpeta para verificar si se encuentra dascargado el paquete de archivos que usaré

In [0]:
cd ..

/content/gdrive/My Drive/Colab Notebooks


In [0]:
ls

'Copia de Copia de SisDist_S1_L1.ipynb'                Examen_Final.ipynb
'Copia de Copy of MCA_SistDist_Ex1'                    Sesion2_Sist_dist2.ipynb
'Copia de Hola, Colaboratory'                          SisDist2_Ex1.ipynb
'Copia de Hola, Colaboratory (1)'                     'SisDist2_S2 (1).ipynb'
'Copia de Hola, Colaboratory (2)'                      SisDist2_S2.ipynb
'Copia de Sesion2_Sist_dist2.ipynb'                    SisDist2_sesion2.ipynb
'Copia de SisDist_S1_L1.ipynb'                         SisDist_S1_L1.ipynb
'Copia de SisDist_S5.ipynb'                            SisDist_S5.ipynb
'Copia de SistDist_S4 (1).ipynb'                       SistDist2/
'Copia de SistDist_S4.ipynb'                           SistDist2_Exam1.ipynb
'Copia de SistDist_S6'                                 Sist_Dist2_Examen1.ipynb
'Copia de Te damos la bienvenida a Colaboratory'       SistDist_S4.ipynb
'Copia de Te damos la bienvenida a Colaboratory (1)'   SistDist_S6.ipynb
'Copia de Te damos la bienv

el paquete trip_data.7z si se encuentra en mi drive de google. Vamos a leer el conjunto de archivos. Sin embargo aún nos faltan descomprimir algunos archivos, lo haremos en la carpeta ExamData

In [0]:
!7z e trip_data.7z -o/content/gdrive/My\ Drive/Colab\ Notebooks/ExamData trip_data_2.csv


7-Zip [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.UTF-8,Utf16=on,HugeFiles=on,64 bits,2 CPUs Intel(R) Xeon(R) CPU @ 2.30GHz (306F0),ASM,AES-NI)

Scanning the drive for archives:
  0M Scan         1 file, 4102781969 bytes (3913 MiB)

Extracting archive: trip_data.7z
--
Path = trip_data.7z
Type = 7z
Physical Size = 4102781969
Headers Size = 433
Method = LZMA:24
Solid = -
Blocks = 12

  0%      0% - trip_data_2.csv                        1% - trip_data_2.csv                        2% - trip_data_2.csv                        3% - trip_data_2.csv                        4% - trip_data_2.csv                        5% - trip_data_2.csv                      

In [0]:
!7z e trip_data.7z -o/content/gdrive/My\ Drive/Colab\ Notebooks/ExamData trip_data_3.csv

In [0]:
!7z e trip_data.7z -o/content/gdrive/My\ Drive/Colab\ Notebooks/ExamData trip_data_4.csv

In [0]:
cd ExamData

/content/gdrive/My Drive/Colab Notebooks/ExamData


In [0]:
ls

trip_data_1.csv   trip_data_3.csv*  Untitled0.ipynb
trip_data_2.csv*  trip_data_4.csv*


Para  responder a las preguntas del punto 6, primero leemos los archivos

In [0]:
archivos = da.read_csv("*.csv")
prom = archivos["trip_distance"].mean()
inicio = time.time()
promedio = prom.compute()
fin = time.time()
print("El tiempo que tarda en obtener el promedio de la distancia de viaje: {} seg".format(fin - inicio))

/usr/local/lib/python3.6/dist-packages/dask/core.py:118: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  args2 = [_execute_task(a, cache) for a in args]


El tiempo que tarda en obtener el promedio de la distancia de viaje: 356.3200843334198 seg


In [0]:

prom = archivos["trip_time_in_secs"].mean()
inicio = time.time()
promedio = prom.compute()
fin = time.time()
print("El tiempo que tarda en obtener el promedio de la duracion de viaje: {} seg".format(fin - inicio))

/usr/local/lib/python3.6/dist-packages/dask/core.py:118: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  args2 = [_execute_task(a, cache) for a in args]


El tiempo que tarda en obtener el promedio de la duracion de viaje: 379.7057626247406 seg


In [0]:
# Convert 'pickup_datetime' column to datetime objects
archivos["pickup_datetime"] = da.to_datetime( archivos["pickup_datetime"] )
# Convert 'dropoff_datetime' column to datetime objects
archivos["dropoff_datetime"] = da.to_datetime( archivos["dropoff_datetime"] )
# Make column 'duracion'
archivos["duracion"] = archivos["dropoff_datetime"] - archivos["pickup_datetime"]
# Construct column 'duracion'
archivos["duracion"] = archivos["duracion"].dt.total_seconds()
out = archivos[(archivos["duracion"] != archivos["trip_time_in_secs"] )]
print("Hay {} datos diferentes".format(archivos.compute().shape[0] - out.compute().shape[0]))

/usr/local/lib/python3.6/dist-packages/dask/core.py:118: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  args2 = [_execute_task(a, cache) for a in args]


Para responder la ultima preunta del punto 6 usaré delayed

In [0]:
from dask import delayed

In [0]:
plantilla = 'trip_data_{:d}.csv'

archivos = [plantilla.format(k) for k in range(1,5)] # comprension de lista
archivos

['trip_data_1.csv', 'trip_data_2.csv', 'trip_data_3.csv', 'trip_data_4.csv']

In [0]:
@delayed
def cuenta_viajes_largos(df):
  da['duracion'] = (da.dropoff_datetime - da.pickup_datetime).dt.seconds
  es_viaje_largo = da.duracion > 1200 # viajes que duran mas de 20 minutos
  result_dict = {'n_long':[sum(es_viaje_largo)],
                'n_total':[len(da)]}
  return pd.DataFrame(result_dict)

In [0]:
@delayed
def abre_archivo(nombre):
  return pd.read_csv(nombre, parse_dates=[1,2])

In [0]:
totales = [cuenta_viajes_largos(abre_archivo(archivo)) for archivo in archivos]
print(totales)

[Delayed('cuenta_viajes_largos-b4c755ae-1f34-43cc-bc23-a2d9af157410'), Delayed('cuenta_viajes_largos-62a9bf96-81f1-45d6-b2ee-fbe8792800fb'), Delayed('cuenta_viajes_largos-39d7993c-6b5d-4807-b0f7-ffdc8606cfb5'), Delayed('cuenta_viajes_largos-c743fa08-8f2d-4dd8-8a2e-98692379e855')]


In [0]:
totales_semi = sum(totales)

7. Contar el numero de viajes largos en cada archivo

In [0]:
totales_semi = totales_semi.compute()

8. De los viajes largos, identificar el numero de taxis diferentes (la columna medallion contiene un numeroque identificada a cada uno de los vehiculos). Que vehiculos son los que mas viajes realizan en cada mes? Son el mismo vehiculo?

'My Drive'/
